03-rec-poly/homework-03/hw-03.hs

```diff
@@ -0,0 +1,91 @@
+For example:
+== ["ABCD", "BD", "C", "D"]
+== ["hello!", "el!", "l!", "l", "o", "!"]
+== [[1]]
+skips "ABCD"
+skips "hello!"
+skips [1]
+skips [True,False] == [[True,False], [False]]
+skips []           == []
+Note that the output should be the same length as the input.
+-}
+--Exercise 1
+skips :: forall a. [a] -> [[a]]
+skips xs = map (\(idx, _) -> filterIdx idx ps) ps
+  where
+    ps :: [(Int, a)]
+    ps = zip [1 .. ] xs
+
+    filterIdx :: Int -> [(Int, a)] -> [a]
+    filterIdx n ls = map snd $ filter (\(idx, _) -> idx `mod` n == 0) ls
+
+
+{-
+Exercise 2 Local maxima
+A local maximum of a list is an element of the list which is strictly
+greater than both the elements immediately before and after it. For example, in the list [2,3,4,1,5], the only local maximum is 4, since it is greater than the elements immediately before and after it (3 and 1). 5 is not a local maximum since there is no element that comes after it.
+Write a function
+localMaxima :: [Integer] -> [Integer]
+which finds all the local maxima in the input list and returns them in order. For example:
+ localMaxima [2,9,5,6,1] == [9,6]
+ localMaxima [2,3,4,1,5] == [4]
+ localMaxima [1,2,3,4,5] == []
+-}
+--Exercise 2
+localMaxima :: [Integer] -> [Integer]
+localMaxima (x : y : z : zs)
+    | y > x && z > y = y : localMaxima (z : zs)
+    | otherwise      = localMaxima (y : z : zs)
+localMaxima _ = []
+
+
+{-
+Exercise 3 Histogram
+For this task, write a function
+histogram :: [Integer] -> String
+which takes as input a list of Integers between 0 and 9 (inclusive), and outputs a vertical histogram showing how many of each number were in the input list. You may assume that the input list does not contain any numbers less than zero or greater than 9 (that is, it does not matter what your function does if the input does contain such numbers). Your output must exactly match the output shown in the examples below.
+cis 194: homework 3 3
+histogram [1,1,1,5] ==
+*
+* **
+ ==========
+ 0123456789
+ histogram [1,4,5,4,6,6,3,4,2,4,9] ==
+*
+* **
+  ******  *
+ ==========
+ 0123456789
+Important note: If you type something like histogram [3,5] at the ghci prompt, you should see something like this:
+ "   * *    \n==========\n0123456789\n"
+This is a textual representation of the String output, including \n escape sequences to indicate newline characters. To actually visualize the histogram as in the examples above, use putStr, for example, putStr (histogram [3,5]).
+-}
+--Exercise 3
+histogram :: [Integer] -> String
+histogram xs = unlines (map (\row -> genLine freqMap row) [mostFreq, mostFreq - 1 .. 1]) ++ "==========\n0123456789\n"
+  where
+    getFreq :: [Integer] -> [Int] 
+    getFreq xs = map (\nums -> length nums - 1) . group . sort $ xs ++ [0 .. 9]
```

Comment:
@jhrcek Hi 😄, could I trouble you and check if you have a cleaner way of doing this with `group` and `sort`? (by cleaner I mean maybe without the use of `[0 .. 9]`)

Reply:
Even your original solution was mostly fine. 10x O(n) is still O(n).
I just wanted to get you thinking about algorithmic complexity.
I like this second solution of yours more than the original :+1: 

Here's couple of more solutions that I can think of.
Most of them require some additional knowledge of libraries, but I think it won't hurt to stretch your imagination :smiley: 

```haskell
$ stack repl --package containers

λ> -- Similar to your solution, but without `++[0..9]`
λ> import Data.List (lookup, group, sort)
λ> import Data.Maybe (fromMaybe)
λ> -- Using head here is safe, because group guarantees that each group is non-empty
λ> numCounts = map (\g -> (head g, length g)) . group . sort
λ> counts = numCounts [1,4,5,4,6,6,3,4,2,4,9]
λ> counts
[(1,1),(2,1),(3,1),(4,4),(5,1),(6,2),(9,1)]
λ> lookup 0 counts
Nothing
λ> lookup 1 counts
Just 1
λ> [fromMaybe 0 (lookup i counts) | i <- [0..9]]
[0,1,1,1,4,1,2,0,0,1]
λ> 
λ> -- A bit more advanced solution using IntMap from containers package
λ> -- https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-IntMap-Strict.html
λ> -- no explicit sort required, also is resilient against invalid numbers in the input list
λ> import qualified Data.IntMap.Strict as M
λ> numCounts2 xs = M.fromListWith (+) $ zip xs (repeat 1)
λ> counts2 = numCounts2 [1,4,5,4,6,6,3,4,2,4,9]
λ> counts2
fromList [(1,1),(2,1),(3,1),(4,4),(5,1),(6,2),(9,1)]
λ> fmap (\i -> M.findWithDefault 0 i counts2) [0..9]
[0,1,1,1,4,1,2,0,0,1]
λ> 
λ> -- Another using GHC.Arr (https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-Arr.html).
λ> -- Most efficient (just single O(n) traversal through the input list, array lookup is O(1))
λ> -- but relies on the input data to be valid (must not contain numbers out of 0 - 9 range)
λ> import GHC.Arr 
λ> numCounts3 xs = accumArray (+) 0 (0,9) $ zip xs (repeat 1)
λ> counts3 = numCounts3 [1,4,5,4,6,6,3,4,2,4,9]
λ> counts3
array (0,9) [(0,0),(1,1),(2,1),(3,1),(4,4),(5,1),(6,2),(7,0),(8,0),(9,1)]
λ> fmap (\i -> counts3 ! i) [0..9]
[0,1,1,1,4,1,2,0,0,1]
```

Reply:
@jhrcek Thank you, this is awesome 👍